integration/test/Test/Connection.hs

```diff
@@ -0,0 +1,403 @@
+  putConnection alice charlie1 "accepted" `waitForResponse` \resp ->
+    resp.status `shouldMatchInt` 200
+
+  -- CHARLIE 2
+
+  -- an incoming connection beyond the limit should make it
+  -- impossible for alice to accept
+  postConnection charlie2 alice `bindResponse` \resp ->
+    resp.status `shouldMatchInt` 201
+
+  putConnection alice charlie2 "accepted" `waitForResponse` \resp -> do
+    resp.json %. "label" `shouldMatch` "connection-limit"
+    resp.status `shouldMatchInt` 403
+
+  -- the status should stay pending
+  getConnection alice charlie2 `bindResponse` \resp -> do
+    resp.json %. "status" `shouldMatch` "pending"
+    resp.status `shouldMatchInt` 200
+
+  -- CHARLIE 5
+
+  -- the remote should be able to accept
+  putConnection charlie5 alice "accepted" `waitForResponse` \resp ->
+    resp.status `shouldMatchInt` 200
+
+  -- the status should change for alice as well
+  getConnection alice charlie5 `waitForResponse` \resp -> do
+    resp.json %. "status" `shouldMatch` "accepted"
+    resp.status `shouldMatchInt` 200
+
+  -- CHARLIE 3
+
+  -- attempting to send a new connection request should also hit the limit
+  postConnection alice charlie3 `waitForResponse` \resp -> do
+    resp.json %. "label" `shouldMatch` "connection-limit"
+    resp.status `shouldMatchInt` 403
+
+  -- CHARLIE 4
+
+  -- blocking should not count towards the connection limit, so after blocking
+  -- charlie 1, we should be able establish another connection
+  putConnection alice charlie1 "blocked" `bindResponse` \resp ->
+    resp.status `shouldMatchInt` 200
+
+  postConnection alice charlie4 `bindResponse` \resp ->
+    resp.status `shouldMatchInt` 201
+
+testNonFederatingRemoteTeam :: HasCallStack => App ()
+testNonFederatingRemoteTeam =
+  withFederatingBackendsAllowDynamic $ \(domainA, domainB, _) -> do
+    sequence_
+      [ createFedConn domainA (FedConn domainB defSearchPolicy Nothing),
+        createFedConn domainB (FedConn domainA defSearchPolicy Nothing)
+      ]
+    void $ updateFedConn domainA domainB (FedConn domainB defSearchPolicy $ Just [])
+    alice <- randomUser domainA def
+    bob <- randomUser domainB def
+    postConnection alice bob `bindResponse` \resp -> do
+      resp.status `shouldMatchInt` 403
+      resp.json %. "label" `shouldMatch` "team-not-federating"
+  where
+    defSearchPolicy = "full_search"
+
+testNonMutualFederationConnectionAttempt :: HasCallStack => App ()
+testNonMutualFederationConnectionAttempt =
+  withFederatingBackendsAllowDynamic $ \(domainA, domainB, _) -> do
+    sequence_
+      [ createFedConn domainA (FedConn domainB defSearchPolicy Nothing),
+        createFedConn domainB (FedConn domainA defSearchPolicy Nothing)
+      ]
+    alice <- randomUser domainA def
+    bob <- randomUser domainB def {API.BrigInternal.team = True}
+
+    -- Alice's backend federates with Bob's team
+    void $ updateFedConn domainA domainB (FedConn domainB defSearchPolicy $ Just [])
```

Comment:
As above I would prefer to create the fedration policy the way we want right away.